### PR TITLE
Added publicIp/port information to host/service

### PR DESCRIPTION
**Pull request by alena1108**
https://github.com/rancher/rancher/issues/2695

With this PR, "service" and "host" objects get one extra API field called "publicEndpoints". Each endpoint has 2 fields:

* port (int)
* ipAddress (string)

Field name is debatable, I would agree with any name having better representation of ip/port combo.

Field gets populated for service/host on every nic.activate/nic.deactivate event for the instance having port exposed. Supported for both service and standalone instances; in standalone instance case only host object will get an update.

This model will be valid even in the future, when we add support for multiple interfaces for hosts + mapping port to a particular interface.

@vincent99 @ibuildthecloud API police, please review

### CODE REVIEW

**Comment by vincent99** [comment]
Description sounds good to me..

**Comment by ibuildthecloud** [comment]
I like it!   Need to review one more time on a screen larger than 4"

File: code/iaas/model/src/main/java/io/cattle/platform/core/addon/IpPortPublicEndpoint.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
+package io.cattle.platform.core.addon;
+
+import io.github.ibuildthecloud.gdapi.annotation.Type;
+
+@Type(list = false)
+public class IpPortPublicEndpoint {
```
Comment (ibuildthecloud) [inline code/iaas/model/src/main/java/io/cattle/platform/core/addon/IpPortPublicEndpoint.java]: Can we change this to PublicEndpoint, as the type ipPortPublicEndpoint will be in the API, ipPort seems unneeded

File: code/iaas/model/src/main/java/io/cattle/platform/core/constants/PortConstants.java [hunk 2]
```diff
@@ -5,4 +5,8 @@
     public static final String KIND_USER = "userPort";
     public static final String KIND_IMAGE = "imagePort";
 
+    public static final String PROCESS_PORT_ACTIVATE = "port.activate";
+    public static final String PROCESS_PORT_DEACTIVATE = "port.deactivate";
+
+    public static final String FIELD_PUBLIC_PORTS = "publicPorts";
```
Comment (ibuildthecloud) [inline code/iaas/model/src/main/java/io/cattle/platform/core/constants/PortConstants.java]: Is this constant used?

File: code/iaas/model/src/main/java/io/cattle/platform/core/dao/impl/HostDaoImpl.java [hunk 3]
```diff
@@ -1,14 +1,28 @@
...
+import io.cattle.platform.lock.LockManager;
+import io.cattle.platform.object.ObjectManager;
 
 import java.util.List;
 
+import javax.inject.Inject;
+
 public class HostDaoImpl extends AbstractJooqDao implements HostDao {
+    @Inject
+    LockManager lockManager;
```
Comment (ibuildthecloud) [inline code/iaas/model/src/main/java/io/cattle/platform/core/dao/impl/HostDaoImpl.java]: Is this used?

File: code/iaas/model/src/main/java/io/cattle/platform/core/dao/impl/InstanceDaoImpl.java [hunk 4]
```diff
@@ -143,4 +149,29 @@ public Instance getInstanceByUuidOrExternalId(Long accountId, String uuid, Strin
                 .fetchInto(InstanceRecord.class);
     }
 
+    @Override
+    public Long getInstanceHostId(long instanceId) {
+        List<? extends InstanceHostMap> maps = mapDao.findNonRemoved(InstanceHostMap.class, Instance.class,
+                instanceId);
+        if (!maps.isEmpty()) {
```
Comment (ibuildthecloud) [inline code/iaas/model/src/main/java/io/cattle/platform/core/dao/impl/InstanceDaoImpl.java]: nit: its easier to read code when the first block of a if statement is the true case.  so in this situation.

```java
if (maps.isEmpty()) {
  return null
}

return maps.get(0).getHostId();
```

File: code/iaas/service-discovery/server/src/main/java/io/cattle/platform/servicediscovery/service/impl/ServiceDiscoveryServiceImpl.java [hunk 5]
```diff
@@ -298,4 +312,51 @@ public boolean isGlobalService(Service service) {
...
+            if (add) {
+                publicEndpoints.add(publicEndpoint);
+            } else {
+                publicEndpoints.remove(publicEndpoint);
+            }
+            DataUtils.getWritableFields(service).put(ServiceDiscoveryConstants.FIELD_PUBLIC_ENDPOINTS,
+                    publicEndpoints);
+            objectManager.persist(service);
+            Map<String, Object> data = new HashMap<>();
+            data.put(ServiceDiscoveryConstants.FIELD_PUBLIC_ENDPOINTS, publicEndpoints);
```
Comment (ibuildthecloud) [inline code/iaas/service-discovery/server/src/main/java/io/cattle/platform/servicediscovery/service/impl/ServiceDiscoveryServiceImpl.java]: We've copied and pasted this code in a couple places now.  Can we make it generic somewhere.  So you can just call `utils.TriggerStateChanged(eventService, resource)`.  And the call to eventService.publish should be deferred.

**Comment by ibuildthecloud** [comment]
@alena1108 Just a couple small comments.  I like how this just keys off of port activate/deactivate.  Very clean, I hope it stays that simple.

File: code/iaas/model/src/main/java/io/cattle/platform/core/addon/IpPortPublicEndpoint.java (hunk 1 above)
Comment (alena1108) [inline code/iaas/model/src/main/java/io/cattle/platform/core/addon/IpPortPublicEndpoint.java]: sure

File: code/iaas/model/src/main/java/io/cattle/platform/core/constants/PortConstants.java (hunk 2 above)
Comment (alena1108) [inline code/iaas/model/src/main/java/io/cattle/platform/core/constants/PortConstants.java]: nope, leftovers from the previous design, will remove

File: code/iaas/model/src/main/java/io/cattle/platform/core/dao/impl/HostDaoImpl.java (hunk 3 above)
Comment (alena1108) [inline code/iaas/model/src/main/java/io/cattle/platform/core/dao/impl/HostDaoImpl.java]: will cleanup

File: code/iaas/service-discovery/server/src/main/java/io/cattle/platform/servicediscovery/service/impl/ServiceDiscoveryServiceImpl.java (hunk 5 above)
Comment (alena1108) [inline code/iaas/service-discovery/server/src/main/java/io/cattle/platform/servicediscovery/service/impl/ServiceDiscoveryServiceImpl.java]: sure

**Pull request merged**